services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/extensions/dynamodb/mappingclient/core/DynamoDbAsyncMappedIndex.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.extensions.dynamodb.mappingclient.core;
+
+import static software.amazon.awssdk.extensions.dynamodb.mappingclient.core.Utils.createKeyFromItem;
+
+import java.util.concurrent.CompletableFuture;
+
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.AsyncMappedIndex;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.IndexOperation;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.Key;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.MapperExtension;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.PaginatedIndexOperation;
+import software.amazon.awssdk.extensions.dynamodb.mappingclient.TableSchema;
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
+
+@SdkPublicApi
+@ThreadSafe
+public class DynamoDbAsyncMappedIndex<T> implements AsyncMappedIndex<T> {
+    private final DynamoDbAsyncClient dynamoDbAsyncClient;
+    private final MapperExtension mapperExtension;
+    private final TableSchema<T> tableSchema;
+    private final String tableName;
+    private final String indexName;
+
+    DynamoDbAsyncMappedIndex(DynamoDbAsyncClient dynamoDbAsyncClient,
```

Comment:
It's implementing a public interface. The class itself doesn't need to be public. We should review all the classes in one go as a separate task for surface area, I'd like to defer on this for now.

Reply:
I see, so this is more like a test api? Can we annotate it with `@SdkTestInternalApi` to make it more clear?

Reply:
No, definitely not a test class, it's roughly analogous to DefaultS3Client implementing S3Client. We'll correctly categorize it when we execute the backlog item to audit surface area of all classes.

Reply:
I meant adding the annotation on the ctor if it is only used for testing.

Reply:
Oh no sorry I was looking at the wrong thing. This object is created by a DynamoDbAsyncMappedTable when table(...) is called. The constructor is package protected to prevent direct instantiation from outside of that.

Reply:
Sorry when index(...) is called (not table)